### ISSUE TITLE
Multiple GPU per node could fail silently with KubeflowEnvironment

### ISSUE THREAD

**Issue by RamtinRassoli** [Lightning-AI/lightning#13374]
## 🐛 Bug
If the user tries to submit a ddp job to a Kubeflow env with multi-gpu per node by following [multi-GPU docs](https://pytorch-lightning.readthedocs.io/en/1.4.0/advanced/multi_gpu.html#distributed-data-parallel) and passing the right args (`num_nodes` and `devices`) one of the following would happen:
  - `WORLD_SIZE` and `RANK` are set to total number of processes -> the job gets stuck because `creates_processes_externally=True` doesn't let ddp launch other processes. 
  - `WORLD_SIZE` and `RANK` are set to total number of nodes -> the job starts with only local rank 0 of each node participating in distributed training. 
The major issue here apart from the idle GPUs is that `DDPStrategy` still works correctly and passes the right number of replicas to the distributed sampler:
```python
...
        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
```
So local rank 0 GPUs will get 1/num_processes of the data assuming other (idle) GPUs are processing the rest. All while training is being done only on a subset of the dataset that was assigned to local rank 0 of each node. The user is unaware of this since they assume they passed `devices/gpus` and `num_nodes` to trainer correctly. 

### To Reproduce
N/A (it's how KubeflowEnvironment works)

### Expected behavior
I'm not sure if this is the expected behavior.  I am using Google Vertex AI that runs Kubeflow under the hood. When a Pytorch Lightning job is submitted to Vertex, Pytorch Lightning automatically selects `KubeflowEnvironment` as the cluster environment. 

Please let me know if the expectation is to have a separate cluster environment class for something like VertexAI. I'd be happy to create a PR to add the new Env. But the reason why I decided to report this as a bug are:
1. `KubeflowEnvironment` has two very specific requirements a. nodes with a single GPU and  b. manual creation of the processes. Neither of these requirements are related to or enforced by Kubeflow. The requirements are also not mentioned in the docs and the user wouldn't know this until they look at the code.  
2. The `detect` method of `KubeflowEnvironment` can be used for any Kubernetes env, and the rest of its methods basically implement an especial case of `LightningEnvironment` where the user has to manually run the processes.

**Comment by awaelchli** [comment]
> KubeflowEnvironment has two very specific requirements a. nodes with a single GPU and b. manual creation of the processes. Neither of these requirements are related to or enforced by Kubeflow.

If that's the case then we need to change it. 

We can make the return value of `creates_processes_externally` dependent on how the user launched the process to detect whether we should spawn additional processes or not. This is how it's done in LightningEnvironment to detect whether LOCAL_RANK was passed as an environment variable. How does that sound?


> Pytorch Lightning automatically selects KubeflowEnvironment as the cluster environment.

This works, right?

> Please let me know if the expectation is to have a separate cluster environment class for something like VertexAI. I'd be happy to create a PR to add the new Env.

The main objective for the cluster env is to translate environment variable names into a standard interface. So unless the variables have different names in VertexAI or there are additional ones that need to be taken into consideration, I would suggest not to go for a new environment class.

**Comment by awaelchli** [comment]
@neggert originally contributed this feature, he might have some thoughts too.

**Comment by RamtinRassoli** [comment]
PytorchJob operator sets the WORLD_SIZE to the total number of replicas by default ([here](https://github.com/kubeflow/pytorch-operator/blob/4aeb6503162465766476519339d3285f75ffe03e/pkg/controller.v1/pytorch/pod.go#L269) and [here](https://github.com/kubeflow/community/blob/master/proposals/pytorch-operator-proposal.md#resulting-master)) which is different from what torch and lightning expect. So `KubeflowEnvironment` should let `DDPStrategy` set global_rank /world_size and create processes externally if needed.

Updating the following methods would be enough to make `KubeflowEnvironment` a generic env that's compatible with Trainer args and multi-gpu clusters:
```python
    @property
    def creates_processes_externally(self) -> bool:
        return "LOCAL_RANK" in os.environ

    def global_rank(self) -> int:
        return self._global_rank

    def set_global_rank(self, rank: int) -> None:
        self._global_rank = rank

    def local_rank(self) -> int:
        return int(os.environ.get("LOCAL_RANK", 0))

    def node_rank(self) -> int:
        return int(os.environ["RANK"])
```
That said this would make it very similar to `LightningEnvironment`. Not sure if that's a problem.

**Comment by awaelchli** [comment]
@RamtinRassoli That's great. Could you quickly look over this original PR to check whether we are still covering the use case that @neggert had with PyTorchJob? I'm not an expert on these. 

I don't understand exactly what 

> So KubeflowEnvironment should let DDPStrategy set global_rank /world_size and create processes externally if needed.

means. In the case where processes get launched externally, why would we let the strategy set the ranks differently? Is this about single node vs multi-node like in the regular LightningEnvironment?

**Comment by RamtinRassoli** [comment]
@awaelchli I checked the PR. I believe all use cases are still covered. Basically when `LOCAL_RANK` is not set the class should behave exactly the same as before. Is there any specific part you want me to check?

> means. In the case where processes get launched externally, why would we let the strategy set the ranks differently? Is this about single node vs multi-node like in the regular LightningEnvironment?

I was referring to this method in `DDPStrategy`:  
```python3
    def set_world_ranks(self) -> None:
        if self.cluster_environment is None:
            return
        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
        rank_zero_only.rank = self.cluster_environment.global_rank()
```
My understanding is when cluster_env is not None the strategy sets the global_rank and world_size. But at the moment, this doesn't work for KubeflowEnvironment as the `set_global_rank` is a no-op.

**Comment by awaelchli** [comment]
Okay, sounds good. Since I'm not a user of this cluster environment, I wanted to make sure we still cover the use case of what was implemented originally. Would you be down to send a PR with these changes to the existing KubeflowEnvironment plugin?

**Comment by neggert** [comment]
Hey all, thanks for pointing this out and sorry for the late response. I think there's an assumption built into PyTorchJob that the user will only be running one process per pod. If you want more processes, you can spin up more pods. There's a slight terminology mis-match because Lightning sees each Kubernetes pod as a separate node. So you _can_ use multiple GPUs per node, they just need to be in separate pods. `num_nodes` is really `num_pods` in Kubernetes.

I guess this one-process-per-pod thing is just an assumption, not a hard requirement, so there's no reason that Lightning couldn't spin up multiple processes per pod. Sounds like the proposed changes would allow you to do that. The other option would be to assert that `devices==num_nodes` in Kubeflow.

I've been out of the game for a while, so my memory is a little rusty. Is the idea that `LOCAL_RANK` would be set for a sub-process spun up by Lightning? PyTorchJob itself will never set `LOCAL_RANK`. If that's the case, I think the changes proposed by @RamtinRassoli will accomplish what you want.

**Comment by RamtinRassoli** [comment]
@neggert yes, `LOCAL_RANK` would be set for subprocesses spun up by PL. And what you said about the PyTorchJob's assumption makes sense, it's just that ideally the KubeflowEnv and LightningEnv should interpret `num_nodes` the same way.

@awaelchli I'll send a PR with the proposed changes soon.
Thank you both!

**Comment by RamtinRassoli** [comment]
@neggert yes, LOCAL_RANK would be set for subprocesses spun up by PL. And what you said about the PyTorchJob's assumption makes sense, it's just that ideally the KubeflowEnv and LightningEnv should interpret num_nodes the same way.

@awaelchli I'll send a PR with the proposed changes soon.
Thank you both!

**Comment by awaelchli** [comment]
Sounds good

> @awaelchli I'll send a PR with the proposed changes soon.

Awesome!

**Comment by almarengo** [comment]
Was this issue ever resolved? Was the PR created and merged? I think I'm running into the same issue: running lightning on Vertex AI on 2 nodes with 4 GPU each and the training hangs. Thanks

**Comment by almarengo** [comment]
Thanks for your help. Disabling the the automatic detention did half the job. The training completed but using only 1 of the 2 nodes. It was distributed on 4 GPUs of the node though. Any other trick to fix this other issue? FYI I passed `num_nodes=2` and `devices=4`.

**Comment by awaelchli** [comment]
@almarengo I am not familiar with how Vertex AI uses kubeflow and what the expectations are in how the jobs are launched and configured. Unfortunately, I won't have the capacity to look into it. We would happily accept contributions and proposals from the community if the existing Kubeflow environment should be changed or updated.

**Comment by almarengo** [comment]
@awaelchli thank you! Appreciate it

**Comment by almarengo** [comment]
@awaelchli I had a closer look at Vertex AI logs. It looks like it span only 2 processes (1GPU per node). I took me a while to track down the reason for that but it looks like the code @RamtinRassoli posted above would fix the issue. Right now the `creates_processes_externally` is set to `True` so it skips the multi-node process creation. Also the `global_rank` needs to be setup properly.

The proposed changes (as per above):
```
    @property
    def creates_processes_externally(self) -> bool:
        return "LOCAL_RANK" in os.environ

    def global_rank(self) -> int:
        return self._global_rank

    def set_global_rank(self, rank: int) -> None:
        self._global_rank = rank

    def local_rank(self) -> int:
        return int(os.environ.get("LOCAL_RANK", 0))

    def node_rank(self) -> int:
        return int(os.environ["RANK"])
```

Was a PR ever created for this? I don't seem to find it. Thanks

**Comment by awaelchli** [comment]
@almarengo The modifications you point out are basically what the default LightningEnvironment already does:
https://github.com/Lightning-AI/lightning/blob/4dfc09c2fe5cf4a66b6ec69e4189fe0ac2a92fb8/src/lightning/fabric/plugins/environments/lightning.py#L45-L52

After https://github.com/Lightning-AI/lightning/pull/18137, the KubeflowEnvironment will no longer be selected automatically and so you should get the desired behavior already from the default.

**Comment by almarengo** [comment]
@awaelchli I was using the KubeflowEnvironment because with the default LightningEnvironment the training was hanging endlessly. I was using Kubeflow on Vertes AI so I assumed that I had to switch to the proper cluster environment. 
I went back to figure out why LightningEnvironment was not working and I think I have the answer. 
The default LightningEnvironment is spinning 8 processes correctly (2 nodes with 4 gpus each) but it's assigning the `GLOBAL_RANK` wrong. Instead of assigning 0 to 7 ranks I have 2 sets of 0 to 3 ranks.
Here is where it's assigned by the DPPStrategy:
![image](https://github.com/Lightning-AI/lightning/assets/77657772/db289093-cc72-4abf-82e7-c2146c583e42)
And the node_rank is assigned by the default LightningEnvironment here:
![image](https://github.com/Lightning-AI/lightning/assets/77657772/20a1d34a-8973-436b-8eed-f0a7d3475624)

The problem is that Kubeflow on Vertex AI does not have either of the environmental variables (`GROUP_RANK` and `NODE_RANK`). The env variable for the node is just called `RANK`. 
So both nodes get assigned the default 0 and the global ranks get duplicated.
Does this make sense? Has anyone else experienced this?

**Comment by almarengo** [comment]
Hey @awaelchli just touching base here, haven't heard anything in a while. Is that a known issue or am I doing anything wrong? Thanks a lot

**Comment by awaelchli** [comment]
Hey @almarengo 
For context, these plugins (MPI, KubeFlow etc) are community contributions and maintained by those who are using these environments. I am a SLURM and torchrun boy so have never come in contact with these and wouldn't be able to give good advice. Any fixes I made in the past I relied on someone from the community to confirm (which is awesome). So to answer your question, I don't know if you are doing something wrong. But if Vertex AI doesn't provide the GROUP_RANK or NODE_RANK to the job, then they might have these variables under a different name. 

The one core question with these environments you have to always ask first and find an answer to is this: Does the system (here Vertex AI's launcher) create 1 process per GPU already for you and launches your script into it (like SLURM does) OR is it the users's/application's responsibility to launch these processes on each node? It's one or the other. That is essentially what the bool property `creates_processes_externally` in the Environment class means. Knowing the answer here would be a good start for the remaining investigation.

**Comment by miguelalba96** [comment]
Maybe I am a bit late for the conversation but as far as I see using vertex AI pipelines (which uses kubeflow), runs `CustomJobs` that are launched per node, then I think what you need to create is a custom `ClusterEnvironment` that parses the information the environment variable `CLUSTER_SPEC` [(see this)](https://cloud.google.com/vertex-ai/docs/training/distributed-training#cluster-spec-format), that JSON-like contains the information of the current launched node-job and the rest of the nodes in the cluster, I think that info is enough to populate the env variables:
```
MASTER_ADDR
MASTER_PORT
WORLD_SIZE
LOCAL_RANK
RANK
```

**Comment by lneves-sc** [comment]
hey, @almarengo!  I am running into a very similar problem with VAI. Did you find a solution for this issue?
